<!DOCTYPE html>



  


<html class="theme-next mist use-motion" lang="zh-Hans">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />




  
  
  
  

  
    
    
  

  

  

  

  
    
      
    

    
  

  
    
    
    <link href="//fonts.googleapis.com/css?family=Lato:300,300italic,400,400italic,700,700italic|Microsoft YaHei:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.4" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.4">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.4" color="#222">





  <meta name="keywords" content="bioinfomatics," />





  <link rel="alternate" href="/atom.xml" title="冷洋" type="application/atom+xml" />






<meta name="description" content="生物信息信息学流程搭建工具－snakemake教程第一步：配置环境需要的软件这些软件都可以直接用conda安装1conda install -y snakemake bwa samtools bcftools graphviz pyyaml docutils 下载示例数据12wget https://bitbucket.org/snakemake/snakemake-tutorial/get/v3">
<meta name="keywords" content="bioinfomatics">
<meta property="og:type" content="article">
<meta property="og:title" content="生物信息信息学流程搭建工具－snakemake教程">
<meta property="og:url" content="http://yoursite.com/2018/06/13/snakemake-tutorial/index.html">
<meta property="og:site_name" content="冷洋">
<meta property="og:description" content="生物信息信息学流程搭建工具－snakemake教程第一步：配置环境需要的软件这些软件都可以直接用conda安装1conda install -y snakemake bwa samtools bcftools graphviz pyyaml docutils 下载示例数据12wget https://bitbucket.org/snakemake/snakemake-tutorial/get/v3">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-13-15-42-22-2018613.png">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-05-11-24-06-201865.png">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-05-17-52-05-201865.png">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-10-36-11-201866.png">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-17-05-50-201866.png">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-17-38-37-201866.png">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-23-21-38-201866.png">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-23-23-04-201866.png">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-23-51-44-201866.png">
<meta property="og:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-07-00-06-27-201867.png">
<meta property="og:updated_time" content="2018-06-13T12:21:26.503Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="生物信息信息学流程搭建工具－snakemake教程">
<meta name="twitter:description" content="生物信息信息学流程搭建工具－snakemake教程第一步：配置环境需要的软件这些软件都可以直接用conda安装1conda install -y snakemake bwa samtools bcftools graphviz pyyaml docutils 下载示例数据12wget https://bitbucket.org/snakemake/snakemake-tutorial/get/v3">
<meta name="twitter:image" content="http://p8gdfbtx6.bkt.clouddn.com/2018-06-13-15-42-22-2018613.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '5.1.4',
    sidebar: {"position":"right","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://yoursite.com/2018/06/13/snakemake-tutorial/"/>





  <title>生物信息信息学流程搭建工具－snakemake教程 | 冷洋</title>
  





  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?2c41cab48f9a2cf90154f4b323282f73";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>




</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <div class="container sidebar-position-right page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">冷洋</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      
        
        <li class="menu-item menu-item-sitemap">
          <a href="/sitemap.xml" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-sitemap"></i> <br />
            
            站点地图
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2018/06/13/snakemake-tutorial/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="冷洋">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="冷洋">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">生物信息信息学流程搭建工具－snakemake教程</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-06-13T20:21:26+08:00">
                2018-06-13
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物信息学/" itemprop="url" rel="index">
                    <span itemprop="name">生物信息学</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          
             <span id="/2018/06/13/snakemake-tutorial/" class="leancloud_visitors" data-flag-title="生物信息信息学流程搭建工具－snakemake教程">
               <span class="post-meta-divider">|</span>
               <span class="post-meta-item-icon">
                 <i class="fa fa-eye"></i>
               </span>
               
                 <span class="post-meta-item-text">阅读次数&#58;</span>
               
                 <span class="leancloud-visitors-count"></span>
             </span>
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h1 id="生物信息信息学流程搭建工具－snakemake教程"><a href="#生物信息信息学流程搭建工具－snakemake教程" class="headerlink" title="生物信息信息学流程搭建工具－snakemake教程"></a>生物信息信息学流程搭建工具－snakemake教程</h1><h2 id="第一步：配置环境"><a href="#第一步：配置环境" class="headerlink" title="第一步：配置环境"></a>第一步：配置环境</h2><h3 id="需要的软件"><a href="#需要的软件" class="headerlink" title="需要的软件"></a>需要的软件</h3><p><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-13-15-42-22-2018613.png" alt="2018-06-13-15-42-22-2018613"><br>这些软件都可以直接用<code>conda</code>安装<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install -y snakemake bwa samtools bcftools graphviz pyyaml docutils</span><br></pre></td></tr></table></figure></p>
<h3 id="下载示例数据"><a href="#下载示例数据" class="headerlink" title="下载示例数据"></a>下载示例数据</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wget https://bitbucket.org/snakemake/snakemake-tutorial/get/v3.11.0.tar.bz2</span><br><span class="line">tar -xf v3.11.0.tar.bz2 --strip 1</span><br></pre></td></tr></table></figure>
<p>完成后产生一下三个文件</p>
<div align="center"><br><br><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-05-11-24-06-201865.png" alt="2018-06-05-11-24-06-201865"><br></div>


<h3 id="创建snakemake-tutorial环境"><a href="#创建snakemake-tutorial环境" class="headerlink" title="创建snakemake-tutorial环境"></a>创建snakemake-tutorial环境</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda env create --name snakemake-tutorial --file environment.yaml</span><br></pre></td></tr></table></figure>
<h3 id="激活环境"><a href="#激活环境" class="headerlink" title="激活环境"></a>激活环境</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">source activate snakemake-tutorial</span><br></pre></td></tr></table></figure>
<p>##　第二步：流程构建基本步骤<br>snakemake是通过定义不同的rule来工作了，rule把流程分解为不同的小步骤，这种分解的方式是通过将输入文件转化为输出文件来实现的。snakemake可以通过文件名来自动解决依赖。</p>
<hr>
<p><strong>下面以<code>bwa-samtools-bcftools</code> call variant的过程为列来构建<code>snakemake流程</code></strong></p>
<h3 id="１-将reads-mapping到参考基因组上"><a href="#１-将reads-mapping到参考基因组上" class="headerlink" title="１ 将reads mapping到参考基因组上"></a>１ 将reads mapping到参考基因组上</h3><p>首先用创建<code>Snakefile</code>的文件<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">rule bwa_map:</span><br><span class="line">    input:</span><br><span class="line">        &quot;data/genome.fa&quot;,</span><br><span class="line">        &quot;data/samples/A.fastq&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;mapped_reads/A.bam&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;bwa mem &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure></p>
<p>==注意==: 这里输入文件有两个，需要用<code>,</code>隔开</p>
<p><strong>伪执行一下</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">snakemake -np mapped_reads/A.bam</span><br></pre></td></tr></table></figure></p>
<p><code>-n</code>: 提示snakemake执行 <code>--dryrun</code>也就是假运行，只提示运行的逻辑，但不会实际运行。<br><code>-p</code>: 打印shell命令行<br>这里snakemake根据Snakefile中编写的生成mapped_reads/A.bam的rule，显示具体的执行命令，但未执行程序。这样，我们可以很方便的检查Snakefile中执行的命令与我们设想的是否一致。</p>
<div align="center"><br><br><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-05-17-52-05-201865.png" alt="2018-06-05-17-52-05-201865"><br></div>

<ul>
<li><strong>通配符匹配</strong><br>snakemake允许使用通配符匹配多个文件。在示例数据中只要将<code>A</code>替换成<code>sample</code>即可，另外如果在上一步，已经生成了<code>A.bam</code>，那么在这一步中即使通配符匹配到了<code>A.fastq</code>也不会在重新生成<code>A.bam</code>。因为snakemake默认先检查output中的文件是否存在，如果存在，则不执行，若不存在则按照rule生成。<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">rule bwa_map:</span><br><span class="line">    input:</span><br><span class="line">        &quot;data/genome.fa&quot;,</span><br><span class="line">        &quot;data/samples/&#123;sample&#125;.fastq&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;bwa mem &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure>
</li>
</ul>
<p>dryrun一下<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">snakemake -np mapped_reads/&#123;A,B&#125;.bam</span><br></pre></td></tr></table></figure></p>
<div align="center"><br><br><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-10-36-11-201866.png" alt="2018-06-06-10-36-11-201866"><br></div>

<h3 id="2-对比对到参考基因组上的序列进行排序"><a href="#2-对比对到参考基因组上的序列进行排序" class="headerlink" title="2 对比对到参考基因组上的序列进行排序"></a>2 对比对到参考基因组上的序列进行排序</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">rule samtools_sort:</span><br><span class="line">    input:</span><br><span class="line">        &quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;sorted_reads/&#123;sample&#125;.bam&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;samtools sort -n -O bam &#123;input&#125; &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure>
<h3 id="3-给bam文件创建索引"><a href="#3-给bam文件创建索引" class="headerlink" title="3 给bam文件创建索引"></a>3 给bam文件创建索引</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">rule samtools_index:</span><br><span class="line">    input:</span><br><span class="line">        &quot;sorted_reads/&#123;sample&#125;.bam&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;sorted_reads/&#123;sample&#125;.bam.bai&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;samtools index &#123;input&#125;&quot;</span><br></pre></td></tr></table></figure>
<p>snakemake在用samtools创见索引的时候总是出现报错，而且这个报错信息也没有提供到底哪里出错了。</p>
<div align="center"><br><br><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-17-05-50-201866.png" alt="2018-06-06-17-05-50-201866"><br></div>

<p><strong>最后选择用picard来排序和建立索引</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">rule picard_sort:</span><br><span class="line">	input:</span><br><span class="line">		&quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br><span class="line">	output:</span><br><span class="line">		&quot;sorted_reads/&#123;sample&#125;.sorted.bam&quot;</span><br><span class="line">	shell:</span><br><span class="line">		&quot;java -jar ~/biosoft/picard/picard.jar SortSam &quot;</span><br><span class="line">        &quot;I=&#123;input&#125; O=&#123;output&#125; SORT_ORDER=coordinate&quot;</span><br><span class="line"></span><br><span class="line">rule picard_index:</span><br><span class="line">	input:</span><br><span class="line">		&quot;sorted_reads/&#123;sample&#125;.sorted.bam&quot;</span><br><span class="line">	output:</span><br><span class="line">		&quot;sorted_reads/&#123;sample&#125;.sorted.bam.bai&quot;</span><br><span class="line">	shell:</span><br><span class="line">		&quot;java -jar ~/biosoft/picard/picard.jar BuildBamIndex I=&#123;input&#125; O=&#123;output&#125;&quot;</span><br></pre></td></tr></table></figure></p>
<ul>
<li><strong>生成有向无环图（DAG）</strong><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">snakemake --dag sorted_reads/&#123;A,B&#125;.bam.bai | dot -Tsvg &gt; dag.svg</span><br></pre></td></tr></table></figure>
</li>
</ul>
<div align="center"><br><br><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-17-38-37-201866.png" alt="2018-06-06-17-38-37-201866"><br></div>

<p>这样不管是自己还是别人都可以非常清晰的看出来流程的处理逻辑</p>
<h3 id="5-call-variants"><a href="#5-call-variants" class="headerlink" title="5 call variants"></a>5 call variants</h3><p>这里先介绍一下snakemake的另一个函数<code>expand</code>，这个函数在有多个通配符的情况下非常好用。</p>
<p>在前面的例子中我们使用<code>sorted_reads/{sample}.bam</code>匹配到<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">[&quot;sorted_reads/A.bam&quot;, &quot;sorted_reads/B.bam&quot;]</span><br></pre></td></tr></table></figure></p>
<p>但是我们在处理生物信息学数据时经常会有多个重复的情况。<br>比如: <code>A-1.bam, A-2.bam, B-1.ｂam, B-2.bam，A-1.bam.bai, A-2.bam.bai, B-1.ｂam.bai, B-2.bam.bai....</code> 。而命令是<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">shell commmad A-1.bam A-1.bam.bai</span><br></pre></td></tr></table></figure></p>
<p>那么我们直接写成<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">rule one:</span><br><span class="line">    input:</span><br><span class="line">        bam=&quot;sorted_reads/&#123;sample&#125;-&#123;repicate&#125;.bam&quot;</span><br><span class="line">        bai=&quot;sorted_reads/&#123;sample&#125;-&#123;repicate&#125;.bam.bai&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;dir/orted_reads/&#123;sample&#125;-&#123;repicate&#125;.vcf&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;shell commmad &#123;input.bam&#125; &#123;input.bai&#125;&quot;</span><br></pre></td></tr></table></figure></p>
<p>即可<br>但是这回匹配所有的<code>A,B,C...</code>样本。如果只想匹配其中的一个或者几个怎么办？<br>第一种方法：把除了需要匹配的都删除。<br>第二种方法：expand函数限定匹配类型。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">rule one:</span><br><span class="line">    input:</span><br><span class="line">        bam=eapand(&quot;sorted_reads/&#123;sample&#125;-&#123;repicate&#125;.bam&quot;, sample=[&apos;A&apos;,&apos;C&apos;], repicate=[1,2])</span><br><span class="line">        bai=expand(&quot;sorted_reads/&#123;sample&#125;-&#123;repicate&#125;.bam.bai&quot;,sample=[&apos;A&apos;,&apos;C&apos;], repicate=[1,2])</span><br><span class="line">    output:</span><br><span class="line">        &quot;dir/orted_reads/&#123;sample&#125;-&#123;repicate&#125;.vcf&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;shell commmad &#123;input.bam&#125; &#123;input.bai&#125;&quot;</span><br></pre></td></tr></table></figure>
<p>当然这是expand最简单的使用场景。</p>
<p><strong>下面正式识别变异</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">SAMPLES=[&quot;A&quot;,&quot;B&quot;]</span><br><span class="line">rule bcftools_call_variants:</span><br><span class="line">    input:</span><br><span class="line">        fa=&quot;data/genome.fa&quot;,</span><br><span class="line">        bam=expand(&quot;sorted_reads/&#123;sample&#125;.sorted.bam&quot;, sample=SAMPLES),</span><br><span class="line">        bai=expand(&quot;sorted_reads/&#123;sample&#125;.sorted.bam.bai&quot;, sample=SAMPLES)</span><br><span class="line">    output:</span><br><span class="line">        &quot;calls/all.vcf&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;samtools mpileup -g -f &#123;input.fa&#125; &#123;input.bam&#125; | &quot;</span><br><span class="line">        &quot;bcftools call -mv - &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure></p>
<p>==注意==: 引号全部用<code>&quot;&quot;</code>不要用单引号。<br><strong>伪执行<code>ｄryrun</code>一下</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">snakemake -np calls/all.vcf</span><br></pre></td></tr></table></figure></p>
<div align="center">

<p><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-23-21-38-201866.png" alt="2018-06-06-23-21-38-201866"><br></p></div><br>再来看看逻辑图<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">snakemake --dag calls/all.vcf|dot -Tsvg&gt;dag2.svg</span><br></pre></td></tr></table></figure><p></p>
<div align="center">

<p><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-23-23-04-201866.png" alt="2018-06-06-23-23-04-201866"><br></p></div><p></p>
<h3 id="6-生成报告"><a href="#6-生成报告" class="headerlink" title="6 生成报告"></a>6 生成报告</h3><p>snakemake一个非常重要的==优点==就是可以直接加上<code>run</code>运行python代码，<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">rule report:</span><br><span class="line">    input:</span><br><span class="line">        &quot;calls/all.vcf&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;report.html&quot;</span><br><span class="line">    run:</span><br><span class="line">        from snakemake.utils import report</span><br><span class="line">        with open(input[0]) as vcf:</span><br><span class="line">            n_calls = sum(1 for l in vcf if not l.startswith(&quot;#&quot;))</span><br><span class="line"></span><br><span class="line">        report(&quot;&quot;&quot;</span><br><span class="line">        An example variant calling workflow</span><br><span class="line">        ===================================</span><br><span class="line"></span><br><span class="line">        Reads were mapped to the Yeast</span><br><span class="line">        reference genome and variants were called jointly with</span><br><span class="line">        SAMtools/BCFtools.</span><br><span class="line"></span><br><span class="line">        This resulted in &#123;n_calls&#125; variants (see Table T1_).</span><br><span class="line">        &quot;&quot;&quot;, output[0], T1=input[0])</span><br></pre></td></tr></table></figure></p>
<p>主要介绍一下<code>report</code>函数，他可以直接渲染mardown到html。主要用到三个参数<br>第一个：markdown内容，可以使用｛n_calls｝样式得到前面已经定义过的python变量。<br>第二个：eg:output[0] 定义html存储位置<br>第三个关键字参数：将关键字参数做key,用value替换，如果需要引用可以在关键字后面加上<code>_</code>下划线。</p>
<div align="center"> 

<p><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-06-23-51-44-201866.png" alt="2018-06-06-23-51-44-201866"><br></p></div><p></p>
<h3 id="生成目标文件"><a href="#生成目标文件" class="headerlink" title="生成目标文件"></a>生成目标文件</h3><p>在以上的例子中通过<code>snakemake target_filename</code>的形式运行流程。而在snakemake中如果没有给出<code>target_filename</code>,snakemake会默认将第一个rule当成的<code>target_filename</code>,因此snakemake支持通过识别顶部<code>rule all</code>来生成最终目标文件。</p>
<div align="center">  

<p><img src="http://p8gdfbtx6.bkt.clouddn.com/2018-06-07-00-06-27-201867.png" alt="2018-06-07-00-06-27-201867"><br></p></div><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">snakemake -n</span><br></pre></td></tr></table></figure><p></p>
<p>发现什么也没发生，这是因为前面我们已经完成了生成<code>report.html</code>文件，snakemake检测到了这个文件，所以不会在执行。当然可以通过<code>--forcerun</code>强制再执行一次。</p>
<h2 id="第三步-进阶"><a href="#第三步-进阶" class="headerlink" title="第三步　进阶"></a>第三步　进阶</h2><h3 id="1-设置线程数"><a href="#1-设置线程数" class="headerlink" title="1 设置线程数"></a>1 设置线程数</h3><p><code>threads</code>设置线程数<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">rule all:</span><br><span class="line">    input:</span><br><span class="line">        &quot;report.html&quot;</span><br><span class="line"></span><br><span class="line">rule bwa_map:</span><br><span class="line">    input:</span><br><span class="line">        &quot;data/genome.fa&quot;,</span><br><span class="line">        &quot;data/samples/&#123;sample&#125;.fastq&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br><span class="line">    threads: 4</span><br><span class="line">    shell:</span><br><span class="line">        &quot;bwa mem -t &#123;threads&#125; &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure></p>
<p>在运行时，可以通过<code>--core</code>指定snakemake可以调用的最大核心数量<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">snakemake --core 6</span><br></pre></td></tr></table></figure></p>
<p>在这个列子中，给了核心给snakemake，但是<code>bwa_map</code>只需要４个，而多出来的两个核心snakemake会自动调度用于<code>picard_sort</code>命令，这样就可以保证资源的最大化利用。如果给定的核心数量少于４，那么snakemake会默认<code>bwa_map</code>只能用能取得的核心数量。</p>
<p><strong>设置内存</strong>：在设置硬件资源时，常用到的就是线程数和内存。snakemake的<code>resource</code>参数能用来设置每个rule的可用内存量或者整个流程的可用内存。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">rule all:</span><br><span class="line">    input:</span><br><span class="line">        &quot;report.html&quot;</span><br><span class="line"></span><br><span class="line">rule bwa_map:</span><br><span class="line">    input:</span><br><span class="line">        &quot;data/genome.fa&quot;,</span><br><span class="line">        &quot;data/samples/&#123;sample&#125;.fastq&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br><span class="line">    threads: 4</span><br><span class="line">    resource:</span><br><span class="line">        mem_mb=100</span><br><span class="line">    shell:</span><br><span class="line">        &quot;bwa mem -t &#123;threads&#125; &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure></p>
<h3 id="２配置文件"><a href="#２配置文件" class="headerlink" title="２配置文件"></a>２配置文件</h3><p>在以上的示例中都是在snakefile内部以python list来给出样本信息，但是大多数情况下我们希望能定制化信息，从而用同一套流程能处理不同情况的数据。snakemake提供<code>configfile</code>配置文件来满足这种需求。配置文件可以用<a href="http://www.w3school.com.cn/json/json_syntax.asp" target="_blank" rel="noopener">JOSN</a>或者<a href="http://www.ruanyifeng.com/blog/2016/07/yaml.html" target="_blank" rel="noopener">YAML</a>编写。</p>
<p>编辑<code>config.yaml</code>内容如下：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">samples:</span><br><span class="line">    A: data/samples/A.fastq</span><br><span class="line">    B: data/samples/B.fastq</span><br></pre></td></tr></table></figure></p>
<p>在运行时snakemake自动解析成一个名为<code>config</code>的字典<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">config=&#123;&apos;sample&apos;: &#123;&apos;A&apos;: &apos;data/samples/A.fastq&apos;, &apos;B&apos;: &apos;data/samples/B.fastq&apos;&#125;&#125;</span><br></pre></td></tr></table></figure></p>
<p>修改后代码如下<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">rule bcftools_call_variants:</span><br><span class="line">    input:</span><br><span class="line">        fa=&quot;data/genome.fa&quot;,</span><br><span class="line">        bam=expand(&quot;sorted_reads/&#123;sample&#125;.sorted.bam&quot;, sample=config[&quot;samples&quot;]),</span><br><span class="line">        bai=expand(&quot;sorted_reads/&#123;sample&#125;.sorted.bam.bai&quot;, sample=config[&quot;samples&quot;])</span><br><span class="line">    output:</span><br><span class="line">        &quot;calls/all.vcf&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;samtools mpileup -g -f &#123;input.fa&#125; &#123;input.bam&#125; | &quot;</span><br><span class="line">        &quot;bcftools call -mv - &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure></p>
<p><strong>其中<code>sample</code>返回的是<code>config[&quot;samples&quot;]</code>这个字典的<code>key</code>即smaple=[‘A’, ‘B’] 这一点非常难以理解,因此我个人在写配置文件的时候基本不会用到snakemake配置文件你这个功能。因为我们完全可以自己定义自己的配置文件格式，然后在rule外用python 读取就可以了</strong></p>
<h3 id="4-设定参数"><a href="#4-设定参数" class="headerlink" title="4 设定参数"></a>4 设定参数</h3><p>很多时候<code>shell commands</code>不仅仅需要input和output，不如在用<code>GATK</code>call variants的时候要求必须有<code>RG</code>，可以在bwa_map这一步加上。<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">samples=[&apos;A&apos;, &apos;B&apos;, &apos;C&apos;]</span><br><span class="line"></span><br><span class="line">rule bwa_map:</span><br><span class="line">    input:</span><br><span class="line">        &quot;data/genome.fa&quot;,</span><br><span class="line">        &quot;data/samples/&#123;sample&#125;.fastq&quot;</span><br><span class="line">    output:</span><br><span class="line">        &quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br><span class="line">    params:</span><br><span class="line">        rg=&quot;@RG\\tID:&#123;sample&#125;\\tSM:&#123;sample&#125;&quot;</span><br><span class="line">    threads: 4</span><br><span class="line">    shell:</span><br><span class="line">        &quot;bwa mem -R &apos;&#123;params.rg&#125;&apos; -t &#123;threads&#125;  &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></pre></td></tr></table></figure></p>
<h3 id="5-日志文件-临时文件-受保护文件"><a href="#5-日志文件-临时文件-受保护文件" class="headerlink" title="5 日志文件 临时文件　受保护文件"></a>5 日志文件 临时文件　受保护文件</h3><p><code>log</code>:参数指定日志问价记录<br><code>temp</code>: 制定临时文件，一旦后面所有用到临时文件作为输入的rule完成就自动删除来节省磁盘空间<br><code>protected</code>: 写保护文件，防止误删除（:smile: rm -rf /* 从删库到跑路）<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">samples=[&apos;A&apos;, &apos;B&apos;, &apos;C&apos;]</span><br><span class="line"></span><br><span class="line">rule bwa_map:</span><br><span class="line">    input:</span><br><span class="line">        &quot;data/genome.fa&quot;,</span><br><span class="line">        &quot;data/samples/&#123;sample&#125;.fastq&quot;</span><br><span class="line">    output:</span><br><span class="line">        temp(&quot;mapped_reads/&#123;sample&#125;.bam&quot;)</span><br><span class="line">       #protected(&quot;mapped_reads/&#123;sample&#125;.bam&quot;) </span><br><span class="line">    params:</span><br><span class="line">        rg=&quot;@RG\\tID:&#123;sample&#125;\\tSM:&#123;sample&#125;&quot;</span><br><span class="line">    threads: 4</span><br><span class="line">    log:</span><br><span class="line">        &quot;logs/bwa_map/&#123;sample&#125;.log&quot;</span><br><span class="line">    shell:</span><br><span class="line">        &quot;(bwa mem -R &apos;&#123;params.rg&#125;&apos; -t &#123;threads&#125;  &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;) 2&gt;&#123;logs&#125;&quot;</span><br></pre></td></tr></table></figure></p>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/bioinfomatics/" rel="tag"># bioinfomatics</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2018/06/13/zsh 插件推荐/" rel="prev" title="zsh 插件推荐">
                zsh 插件推荐 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">冷洋</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">2</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">2</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">2</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          
            <div class="feed-link motion-element">
              <a href="/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#生物信息信息学流程搭建工具－snakemake教程"><span class="nav-number">1.</span> <span class="nav-text">生物信息信息学流程搭建工具－snakemake教程</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#第一步：配置环境"><span class="nav-number">1.1.</span> <span class="nav-text">第一步：配置环境</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#需要的软件"><span class="nav-number">1.1.1.</span> <span class="nav-text">需要的软件</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#下载示例数据"><span class="nav-number">1.1.2.</span> <span class="nav-text">下载示例数据</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#创建snakemake-tutorial环境"><span class="nav-number">1.1.3.</span> <span class="nav-text">创建snakemake-tutorial环境</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#激活环境"><span class="nav-number">1.1.4.</span> <span class="nav-text">激活环境</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#１-将reads-mapping到参考基因组上"><span class="nav-number">1.1.5.</span> <span class="nav-text">１ 将reads mapping到参考基因组上</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-对比对到参考基因组上的序列进行排序"><span class="nav-number">1.1.6.</span> <span class="nav-text">2 对比对到参考基因组上的序列进行排序</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-给bam文件创建索引"><span class="nav-number">1.1.7.</span> <span class="nav-text">3 给bam文件创建索引</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-call-variants"><span class="nav-number">1.1.8.</span> <span class="nav-text">5 call variants</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#6-生成报告"><span class="nav-number">1.1.9.</span> <span class="nav-text">6 生成报告</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#生成目标文件"><span class="nav-number">1.1.10.</span> <span class="nav-text">生成目标文件</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#第三步-进阶"><span class="nav-number">1.2.</span> <span class="nav-text">第三步　进阶</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-设置线程数"><span class="nav-number">1.2.1.</span> <span class="nav-text">1 设置线程数</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#２配置文件"><span class="nav-number">1.2.2.</span> <span class="nav-text">２配置文件</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#4-设定参数"><span class="nav-number">1.2.3.</span> <span class="nav-text">4 设定参数</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-日志文件-临时文件-受保护文件"><span class="nav-number">1.2.4.</span> <span class="nav-text">5 日志文件 临时文件　受保护文件</span></a></li></ol></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">冷洋</span>

  
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Mist</a> v5.1.4</div>




        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  


  











  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>
  

  
  
    <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>
  

  
  
    <script type="text/javascript" src="/lib/canvas-nest/canvas-nest.min.js"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.4"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.4"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.4"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.4"></script>



  


  




	





  





  












  





  

  
  <script src="https://cdn1.lncld.net/static/js/av-core-mini-0.6.4.js"></script>
  <script>AV.initialize("Lji3k2Sno8J07yoLyuipgF4d-gzGzoHsz", "XRxAPD8LzbskF1uPu0CoD8NV");</script>
  <script>
    function showTime(Counter) {
      var query = new AV.Query(Counter);
      var entries = [];
      var $visitors = $(".leancloud_visitors");

      $visitors.each(function () {
        entries.push( $(this).attr("id").trim() );
      });

      query.containedIn('url', entries);
      query.find()
        .done(function (results) {
          var COUNT_CONTAINER_REF = '.leancloud-visitors-count';

          if (results.length === 0) {
            $visitors.find(COUNT_CONTAINER_REF).text(0);
            return;
          }

          for (var i = 0; i < results.length; i++) {
            var item = results[i];
            var url = item.get('url');
            var time = item.get('time');
            var element = document.getElementById(url);

            $(element).find(COUNT_CONTAINER_REF).text(time);
          }
          for(var i = 0; i < entries.length; i++) {
            var url = entries[i];
            var element = document.getElementById(url);
            var countSpan = $(element).find(COUNT_CONTAINER_REF);
            if( countSpan.text() == '') {
              countSpan.text(0);
            }
          }
        })
        .fail(function (object, error) {
          console.log("Error: " + error.code + " " + error.message);
        });
    }

    function addCount(Counter) {
      var $visitors = $(".leancloud_visitors");
      var url = $visitors.attr('id').trim();
      var title = $visitors.attr('data-flag-title').trim();
      var query = new AV.Query(Counter);

      query.equalTo("url", url);
      query.find({
        success: function(results) {
          if (results.length > 0) {
            var counter = results[0];
            counter.fetchWhenSave(true);
            counter.increment("time");
            counter.save(null, {
              success: function(counter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(counter.get('time'));
              },
              error: function(counter, error) {
                console.log('Failed to save Visitor num, with error message: ' + error.message);
              }
            });
          } else {
            var newcounter = new Counter();
            /* Set ACL */
            var acl = new AV.ACL();
            acl.setPublicReadAccess(true);
            acl.setPublicWriteAccess(true);
            newcounter.setACL(acl);
            /* End Set ACL */
            newcounter.set("title", title);
            newcounter.set("url", url);
            newcounter.set("time", 1);
            newcounter.save(null, {
              success: function(newcounter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(newcounter.get('time'));
              },
              error: function(newcounter, error) {
                console.log('Failed to create');
              }
            });
          }
        },
        error: function(error) {
          console.log('Error:' + error.code + " " + error.message);
        }
      });
    }

    $(function() {
      var Counter = AV.Object.extend("Counter");
      if ($('.leancloud_visitors').length == 1) {
        addCount(Counter);
      } else if ($('.post-title-link').length > 1) {
        showTime(Counter);
      }
    });
  </script>



  

  

  
  

  

  

  

</body>
</html>
